Expand Parameter Validation Implicit VU Coverage

mark-lunarg opened the pull request:
This patch series:

- adds implicit VU coverage for structs _only_ appearing in pNext chains
- added VUIDs for validation of any struct or substruct whose contents are validated
- added correct VUIDs for string integrity checks
- added VUIDs for stype checks for arrays of structs

This increases implicit VU coverage by 250+ checks.  

Still around 400 or so to be picked up.

mark-lunarg commented:
Fixed the clang-format issue locally.

tobine commented on scripts/parameter_validation_generator.py:
```diff
@@ -500,6 +544,16 @@ def genStruct(self, typeinfo, typeName, alias):
+        # pNext members. Everything else will be overwritten by the callee.
+        if typeinfo.elem.attrib.get('returnedonly') is not None:
+            if typeName == 'VkPhysicalDeviceGroupProperties':
+                stop = 'here'
```
what do these `stop = 'here'` lines do?

mark-lunarg replied:
This is already gone...

tobine commented on layers/parameter_validation.h:
```diff
@@ -176,6 +176,19 @@ struct LogMiscParams {
+static bool ValidateSubstructure(debug_report_data *report_data, const char *api_name, const char *struct_name,
+                                 const std::string &substructure_vuid) {
+    bool skip = false;
+    if (false) {
```
`"if (false)"`? Wazzup?

mark-lunarg replied:
So.  It was intended that this message would never actually be output to the user.  The spec is kinda weird here -- Each structure has an implicit VU that says 'the structure must be a valid VkXxx structrure', but in addition, all members of that structure get their own VUs.  We could output this message, but than any implicit check would always trigger two errors.  I thought this was better, but am open to suggestions.

tobine replied:
Can you give an example of two overlapping VUs that would be flagged?
For structures I always see the first implicit VU of "sType must be VK_STRUCTURE_TYPE_..."
Are the duplicate errors only a concern when a pNext chain struct has wrong type. If that's the case the conflict still isn't clear to me on just a bit of thought.
The sType and pNext errors are independent to me so I think I'm missing something here.

mark-lunarg replied:
Sure, in CreateInstance, the first implicit VU is "**pCreateInfo must be a valid pointer to a valid VkInstanceCreateInfo structure**" (**VUID-vkCreateInstance-pCreateInfo-parameter**).  But, that same **VkInstanceCreateInfo** structure has the following implicit VUIDs:

* sType must be VK_STRUCTURE_TYPE_INSTANCE_CREATE_INFO
* Each pNext member of any structure (including this one) in the pNext chain must be either NULL or a pointer to a valid instance of VkDebugReportCallbackCreateInfoEXT, VkDebugUtilsMessengerCreateInfoEXT, or VkValidationFlagsEXT
* Each sType member in the pNext chain must be unique
* flags must be 0
* If pApplicationInfo is not NULL, pApplicationInfo must be a valid pointer to a valid VkApplicationInfo structure
* If enabledLayerCount is not 0, ppEnabledLayerNames must be a valid pointer to an array of enabledLayerCount null-terminated UTF-8 strings
* If enabledExtensionCount is not 0, ppEnabledExtensionNames must be a valid pointer to an array of enabledExtensionCount null-terminated UTF-8 strings

Theoretically, **any** failure among those checks should ALSO trigger the CreateInstance VUID failure.

tobine replied:
Ok, that makes more sense. For the first VU my opinion is that we should just flag it if the pCreateInfo ptr is NULL. For all other cases just flag the errors for the specific struct params.

I agree that a strict adherence to VU language would require flagging the first error for any bad struct param, but if I'm a user the redundancy doesn't help me; I only need the more specific error to fix my issue.

tobine commented on scripts/vuid_mapping.py:
```diff
@@ -1168,6 +1173,40 @@
+'depthStencil' : 541,
+'pTessellationState' : 542,
+'pViewportState' : 543,
+'pMultisampleState' : 543,
```
543 duplicated and 545 skipped.

mark-lunarg replied:
Thanks!

tobine commented:
Also, tests? Don't need tests for all cases but perhaps at least 4 new tests to hit one each of the four main classes of checks noted in your PR description?

mark-lunarg commented:
@tobine, moved the fake substructure checks into an existing struct check, which now returns the substruct VUID only if the pointer is null, as you suggested.  This is also much cleaner.  Made the other changes as well, and added four new tests.

tobine commented on scripts/parameter_validation_generator.py:
```diff
@@ -493,6 +534,14 @@ def genStruct(self, typeinfo, typeName, alias):
                                                 condition=conditions[name] if conditions and name in conditions else None,
                                                 cdecl=cdecl))
+        # If this struct extends another, keep its name in list for further processing
+        if (typeinfo.elem.attrib.get('structextends') is not None):
```
parens around non-compound "if" clause are not common in the code, see line 542 below

tobine commented on tests/layer_validation_tests.cpp:
```diff
@@ -1348,6 +1353,74 @@ TEST_F(VkLayerTest, RequiredParameter) {
+        }
+    }
+    // Create a device passing in a bad PdevFeatures2 value
+    auto indexingFeatures = lvl_init_struct<VkPhysicalDeviceDescriptorIndexingFeaturesEXT>();
```
camelCase slipped through
